R/helpers.R

```diff
@@ -698,10 +678,191 @@ h_is_positive_definite <- function(x, tolerance = 1e-08) {
+#' @param digits (`count`)\cr a desired number of significant digits for
+#'   for numbers used in JAGS input, see [`formatC`].
+#' @return Nothing, but as a side effect, the model file is written.
+#'
+#' @example examples/helpers-write_model.R
+#' @export
+#'
+h_write_model <- function(model, file = "model.jags", digits = 5) {
+  assert_function(model)
+  assert_string(file)
+  assert_count(digits)
+
+  # Replace scientific notation.
+  model_sci_replaced <- h_rapply(
+    x = body(model),
+    fun = h_format_number,
+    classes = c("integer", "numeric"),
+    digits = digits,
+    prefix = "TEMP_NUM_PREF_",
+    suffix = "_TEMP_NUM_SUF"
+  )
+  # Transform `model` body into character vector.
+  model_text <- deparse(model_sci_replaced, control = NULL)
+  model_text <- gsub("\"TEMP_NUM_PREF_|_TEMP_NUM_SUF\"", "", model_text)
+  model_text <- gsub("%_% ", "", model_text)
+  model_text <- c("model", model_text)
+
+  writeLines(model_text, con = file)
+}
+
+#' Recursively Apply a Function to a List
+#'
+#' @description `r lifecycle::badge("experimental")`
+#'
+#' This helper function is conceptually quite similar to [`base::rapply`]
+#' function with two major differences (see the note for more details).
+#' It recursively iterates through a "list-like" object and it checks whether
+#' an element is of a given class. If it so, then it replaces that element by
+#' the result of an execution of a given function. Otherwise, and if the element
+#' is of length greater than 1 (i.e. not a scalar), it replaces that element by
+#' the result of `h_rapply`, recursively called for that element. In the
+#' remaining case, that is, the element is not of a given class and is a scalar,
+#' then that element remains unchanged.
+#'
+#' @note This helper function is conceptually similar the same as `rapply`
+#'   function. However, it differs from `rapply` in two major ways. First, the
+#'   `h_rapply` is not limited to objects of type `list` or `expression` only.
+#'   It can be any "list-like" object of any type for which `[[` operator is
+#'   defined. This can be, for example, an object of type `language`, often
+#'   obtained from the `body` function. The second difference is that the
+#'   flexibility of `rapply` on `how` the result is structured is not available
+#'   with `h_rapply` for the user. That is, with `h_rapply` each element of `x`,
+#'   which has a class included in `classes`, is replaced by the result of
+#'   applying `fun` to the element.
+#'
+#' @param x \cr any "list-like" object for which `[[` operator is defined.
+#' @param fun (`function`)\cr a function of one "principal" argument, passing
+#'   further arguments via `...`.
+#' @param classes (`character`)\cr class names.
+#' @param ... further arguments passed to function `fun`.
+#'
+#' @return "list-like" object of similar structure as `x`.
+#'
+#' @example examples/helpers-rapply.R
+#' @export
+#'
+h_rapply <- function(x, fun, classes, ...) {
+  assert_function(fun)
```

Comment:
maybe assert here that `x[[1]]` works?

Reply:
Thanks @danielinteractive I am not so sure what would be the best here. The most reasonable solution which comes to my mind is: `assert_number(invisible(x[[1]]), 1)`. But this is actually obsolete since if x is not subsettable, the error will be thrown from `invisible`, not from the `assert_number` (so the error message will not be that assertion failed). But actually, if `x` is not subsettable, it will anyway fail with the same error at the second line, i.e. ` if (class(x[[i]]) %in% classes)`

Reply:
probably we just get a nicer error and user understands better if we check something simple here, e.g. `assert_true(!is.null(x[[1]]))`? but if too difficult, we leave as is, not that critical

Reply:
yeah, I have tried this approach too, but it will again fail at `x[[1]]` if `x` is not subsettable, without reaching `is.null` fun (see e.g. `assert_true(!is.null(mean[[1]]))` ). Ideally if we have something like `assert_noerror`, but we do not :( So I would leave it as it is.

Reply:
The only one idea which comes to my mind at the moment is to use `tryCatch`.

Reply:
(maybe just have `force(x[[1]])` which evaluates it. if it works, fine, we don't use the result, and otherwise gives clear error here)

Reply:
Thanks @danielinteractive What about `invisible(x[[1]])`, as it does not produce any output if no failure, and it produces the error when there is an error?

Reply:
also fine, `force` is anyway more a readability thing, it just returns the argument